Added DefaultCpuHealthChecker

Bue-von-hon opened the pull request:
Motivation:

Resolves part of #1854

Some users may want to use CPU usage to determine if their instance is healthy.
From now on, Users can perform health checks in the following ways

```java
DefaultCpuHealthChecker cpuHealthChecker = new DefaultCpuHealthChecker(20.0);
cpuHealthChecker.isHealthy();
```

Modifications:

- Added DefaultCpuHealthChecker that checks cpu-usage by OperatingSystemMXBean

Result:

A User can health-check with system-cpu-usage

p.s.
Implemented this by referencing [ProcessorMetrics.java](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/system/ProcessorMetrics.java) in the micrometer core.

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,74 @@
+
+    private final double targetCpuUsage;
+
+    public DefaultCpuHealthChecker(final double targetCpuUsage) {
```
Should we accept arguments for both `targetSystemCpuLoad` and `targetProcessCpuLoad` so that users can configure limits separately?

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    private Method detectMethod(final String name) {
```
```suggestion
    @Nullable
    private Method detectMethod(final String name) {
```

Bue-von-hon replied:
i added!

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    private Class<?> getFirstClassFound(final List<String> classNames) {
```
```suggestion
    @Nullable
    private Class<?> getFirstClassFound(final List<String> classNames) {
```

Bue-von-hon replied:
i added!

jrhee17 reviewed (commented):
Looks good overall! Left some minor comments

Bue-von-hon commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class DefaultCpuHealthChecker implements HealthChecker{
```
I'm struggling with writing tests.😭
To do the test, need to modify the CPU usage of the system and it is very difficult.
I'm considering passing a Method class through the constructor.(Dependency Injection with @VisibleForTesting annotation).
p.s.
I looked at micrometer's ProcessorMetricsTest.java and spring-boot's AvailabilityStateHealthIndicatorTests.java but didn't get much help.😭

jrhee17 replied:
> To do the test, need to modify the CPU usage of the system and it is very difficult.

I don't think this is necessary 😅    I think just verifying that `DefaultCpuHealthChecker` gathers system cpu usage information is enough.
We may expose `targetCpuUsage`, `targetProcessCpuLoad` and check that they are not NaN for example.

Bue-von-hon reviewed (commented):
@jrhee17 Fixed based on a code review!

p.s.
Another PR on the same issue [commented on the ratio-based](https://github.com/line/armeria/pull/4711/files#r1119822052).
I thought it was reasonable so applied it!

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Nullable
+    private Class<?> getFirstClassFound(final List<String> classNames) {
```
```suggestion
    private static Class<?> getFirstClassFound(final List<String> classNames) {
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+        this.targetProcessCpuLoad = (double) processCpuUsage / processCpuIdle;
+        this.operatingSystemBean = ManagementFactory.getOperatingSystemMXBean();
+        this.operatingSystemBeanClass = requireNonNull(getFirstClassFound(OPERATING_SYSTEM_BEAN_CLASS_NAMES));
+        this.systemCpuUsage = requireNonNull(detectMethod("getSystemCpuLoad"));
+        this.processCpuUsage = requireNonNull(detectMethod("getProcessCpuLoad"));
```
It's possible that these methods don't exist.

```suggestion
        this.systemCpuUsage = detectMethod("getSystemCpuLoad");
        this.processCpuUsage = detectMethod("getProcessCpuLoad");
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+     * @param processCpuUsage the process cpu usage
+     * @param processCpuIdle the process cpu idle
+     */
+    public DefaultCpuHealthChecker(final int cpuUsage, final int cpuIdle,
```
No need for parameters to be `final`

```suggestion
    public DefaultCpuHealthChecker(int cpuUsage, int cpuIdle,
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+
+    private final Class<?> operatingSystemBeanClass;
+
+    private final Method systemCpuUsage;
+
+    private final double targetCpuUsage;
+
+    private final Method processCpuUsage;
```
As noted below, it is possible that the method is `null` if the mxbean classes don't exist.
Should we default to the system being healthy in this case? (with a warning log)

```suggestion
    @Nullable
    private final Method systemCpuUsage;

    private final double targetCpuUsage;

    @Nullable
    private final Method processCpuUsage;
```

trustin replied:
> Should we default to the system being healthy in this case? (with a warning log)

Yes

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Nullable
+    private Method detectMethod(final String name) {
+        requireNonNull(name);
```
No need for this since this is a private method and we supply the arguments internally

```suggestion
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+     * @param processCpuUsage the process cpu usage
+     * @param processCpuIdle the process cpu idle
+     */
+    public DefaultCpuHealthChecker(final int cpuUsage, final int cpuIdle,
+                                   final int processCpuUsage, final int processCpuIdle) {
```
Instead of exposing this public constructor, can we use a static factory method to instantiate this checker?

```suggestion
    private DefaultCpuHealthChecker(final int cpuUsage, final int cpuIdle,
                                   final int processCpuUsage, final int processCpuIdle) {
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2015 LINE Corporation
```
```suggestion
 * Copyright 2023 LINE Corporation
```

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * Forked from <a href="https://github.com/micrometer-metrics/micrometer/blob/8339d57bef8689beb8d7a18b429a166f6595f2af/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/system/ProcessorMetrics.java">ProcessorMetrics.java</a> in the micrometer core.
+ */
```
This shouldn't be part of Javadoc. Could you use `//` instead?

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+
+    private double invoke(final Method method) {
+        try {
+            return (double) method.invoke(operatingSystemBean);
```
This method will be frequently invoked whenever `HealthCheckService` receives a request, hence [`MethodHandles`](https://www.baeldung.com/java-method-handles) might be a better choice.

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+ */
+public class DefaultCpuHealthChecker implements HealthChecker {
+
+    private static final List<String> OPERATING_SYSTEM_BEAN_CLASS_NAMES = Arrays.asList(
```
```suggestion
    private static final List<String> OPERATING_SYSTEM_BEAN_CLASS_NAMES = ImmutableList.of(
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,106 @@
+/**
+ * Forked from <a href="https://github.com/micrometer-metrics/micrometer/blob/8339d57bef8689beb8d7a18b429a166f6595f2af/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/system/ProcessorMetrics.java">ProcessorMetrics.java</a> in the micrometer core.
+ */
+public class DefaultCpuHealthChecker implements HealthChecker {
```
If we hide the implementation from the public API, we can get more freedom on changes.
Should we add a factory method of this class to `HealthChecker` and make this package-private?
```
// The simplest one.
HealthChecker.ofCpu(cpuUsage);
// A complex configuation.
HealthChecker.ofCpu(cpuUsage, cpuIdle...);
```

Bue-von-hon reviewed (commented):
@jrhee17 I sent a new commit based on the code review.
Hope you like it🙂

Bue-von-hon commented:
Gentle Ping @trustin @ikhoon @jrhee17

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,122 @@
+ * final boolean healthy = cpuHealthChecker.isHealthy();
+ */
+// Forked from <a href="https://github.com/micrometer-metrics/micrometer/blob/8339d57bef8689beb8d7a18b429a166f6595f2af/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/system/ProcessorMetrics.java">ProcessorMetrics.java</a> in the micrometer core.
+final class DefaultCpuHealthChecker implements HealthChecker {
```
We could rename to `CpuHealthChecker` because we don't have an interface for this class.

Bue-von-hon replied:
renamed it as suggested!

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,122 @@
+        return currentSystemCpuUsage <= targetCpuUsage && currentProcessCpuUsage <= targetProcessCpuLoad;
+    }
+
+    private double invoke(final Method method) {
+        try {
+            final MethodHandles.Lookup lookup = MethodHandles.lookup();
+            final MethodHandle mh = lookup.unreflect(method);
```
Could we keep the `MethodHandle` in a static final field, so we don't have to look up on every invocation? Otherwise, we may have a leak of `MethodHandle`s. (There's a known JDK issue related with this.)

Bue-von-hon replied:
Modified to statically generate or recycle using the cache.

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthChecker.java:
```diff
@@ -0,0 +1,122 @@
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * The Default cpu health checker.<br>
+ * Here's an example of how to use it.<br>
+ * final DefaultCpuHealthChecker cpuHealthChecker = HealthChecker.of(10, 10);<br>
+ * final boolean healthy = cpuHealthChecker.isHealthy();
```
```suggestion
 * A {@link HealthChecker} that reports as unhealthy when the current
 * CPU usage or CPU load exceeds threshold. For example:
 * <pre>{@code
 * final DefaultCpuHealthChecker cpuHealthChecker = HealthChecker.of(10, 10);<br>
 * final boolean healthy = cpuHealthChecker.isHealthy();
 * }</pre>
```

trustin commented on core/src/test/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthCheckerTest.java:
```diff
@@ -0,0 +1,35 @@
+
+import org.junit.jupiter.api.Test;
+
+class DefaultCpuHealthCheckerTest {
```
Could you improve this test so that:

- It actually checks the CPU usage and load (just some range check will be enough)
- It actually checks the healthiness is reported correctly.
  - You could refactor `DefaultCpuHealthChecker` to accept `Supplier`s that provide usage and load, and then inject mock suppliers to test this.

e.g.

```java
class CpuHealthChecker ... {
    CpuHealthChecker(...) {
        this(CpuHealthChecker::cpuUsage, CpuHealthChecker::cpuLoad, ...)
    }

    @VisibleForTesting
    CpuHealthChecker(DoubleSupplier cpuUsageSupplier, DoubleSupplier cpuLoadSupplier, ...) { ... }
}

// In your test:
@Test
void test() {
    final CpuHealthChecker checker = new CpuHealthChecker(() -> 3.0, () -> 4.0, ...);
    // Assertions here
}
```

Bue-von-hon replied:
I've implemented your suggestion by creating a method for testing.
Please take a look and leave a review to see if I did what you suggested.🙏

trustin commented:
@Bue-von-hon Sorry I'm late. Left some comments 🙇

Bue-von-hon commented on core/src/test/java/com/linecorp/armeria/server/healthcheck/DefaultCpuHealthCheckerTest.java:
```diff
@@ -0,0 +1,35 @@
+ * under the License.
+ */
+package com.linecorp.armeria.server.healthcheck;
+
```
consecutive empty lines should be deleted

Bue-von-hon commented:
I've made some improvements that were commented on.
Please take a look. 🙏
@trustin @ikhoon @jrhee17

Bue-von-hon commented:
Gentle Ping @trustin @ikhoon @jrhee17

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,149 @@
+     * @param targetCpuUsage the target cpu usage
+     * @param targetProcessCpuLoad the target process cpu usage
+     */
+    private CpuHealthChecker(double targetCpuUsage, double targetProcessCpuLoad) {
```
Should we rename these to `cpuUsageThreshold` and `processCpuLoadThreshold`? 'target' sounds like we should aim to reach the 'target'.

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,149 @@
+
+    @VisibleForTesting
+    public boolean isHealthy(
+            DoubleSupplier currentSystemCpuUsageSupplier, DoubleSupplier currentProcessCpuUsageSupplier) {
```
These suppliers should be specified in the constructor. For example:
```java

private static final DoubleSuplier defaultSystemCpuUsageSupplier;
private static final DoubleSupplier defaultProcessCpuUsageSupplier;

static {
  .. do reflection, method handle work, etc ..
  defaultSystemCpuUsageSupplier = () -> { ... };
  ..
}

private CpuHealthChecker(double targetCpuUsage, double targetProcessCpuLoad) {
  this(targetCpuUsage, targetProcessCpuLoad, defaultSystemCpuUsageSupplier, defaultProcessCpuUsageSupplier)
}

private CpuHealthChecker(
  double targetCpuUsage, double targetProcessCpuLoad,
  DoubleSupplier systemCpuUsageSupplier, DoubleSupplier processCpuUsageSupplier) {
  ...
  this.systemCpuUsageSupplier = systemCpuUsageSupplier;
  this.processCpuUsageSupplier = ...;
}

@Override
public boolean isHealthy() {
  ...
  final double currentSystemUsage = systemCpuUsageSupplier.get();
  ...
}
```

Bue-von-hon commented:
@trustin I just realized I still have some unresolved comments.😅
I'll try to address them this weekend.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=4673) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=4673) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: Bue-von-hon<br/>:x: taehun.kim<br/><hr/>**taehun.kim** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=4673) it.</sub>

Bue-von-hon commented:
The commit wasn't signed, so it's a week late. 😅
sorry for late, but I've incorporated everything you left in the comments, so please take a look.
@ikhoon @trustin @jrhee17

Bue-von-hon commented:
gentle ping @trustin @ikhoon @jrhee17 @minwoox

jrhee17 commented:
Sorry about the late review, looks almost done!
Can you check the following diff?

```
diff --git a/core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java b/core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java
index 302154226..249999fe0 100644
--- a/core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java
+++ b/core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java
@@ -23,9 +23,11 @@ import java.lang.management.ManagementFactory;
 import java.lang.management.OperatingSystemMXBean;
 import java.lang.reflect.Method;
 import java.util.List;
-import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.DoubleSupplier;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
 
@@ -42,12 +44,17 @@ import com.linecorp.armeria.common.annotation.Nullable;
 // Forked from <a href="https://github.com/micrometer-metrics/micrometer/blob/8339d57bef8689beb8d7a18b429a166f6595f2af/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/system/ProcessorMetrics.java">ProcessorMetrics.java</a> in the micrometer core.
 final class CpuHealthChecker implements HealthChecker {
 
+    private static final Logger logger = LoggerFactory.getLogger(CpuHealthChecker.class);
+
+    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
+
     private static final DoubleSupplier currentSystemCpuUsageSupplier;
 
     private static final DoubleSupplier currentProcessCpuUsageSupplier;
 
     private static final OperatingSystemMXBean operatingSystemBean;
 
+    @Nullable
     private static final Class<?> operatingSystemBeanClass;
 
     private static final List<String> OPERATING_SYSTEM_BEAN_CLASS_NAMES = ImmutableList.of(
@@ -57,16 +64,17 @@ final class CpuHealthChecker implements HealthChecker {
 
     @Nullable
     @VisibleForTesting
-    static final Method systemCpuUsage;
+    static final MethodHandle systemCpuUsage;
 
     @Nullable
     @VisibleForTesting
-    static final Method processCpuUsage;
+    static final MethodHandle processCpuUsage;
 
     static {
         operatingSystemBeanClass = requireNonNull(getFirstClassFound(OPERATING_SYSTEM_BEAN_CLASS_NAMES));
         operatingSystemBean = ManagementFactory.getOperatingSystemMXBean();
-        systemCpuUsage = detectMethod("getSystemCpuLoad");
+        final MethodHandle getCpuLoad = detectMethod("getCpuLoad");
+        systemCpuUsage = getCpuLoad != null ? getCpuLoad : detectMethod("getSystemCpuLoad");
         processCpuUsage = detectMethod("getProcessCpuLoad");
         currentSystemCpuUsageSupplier = () -> invoke(systemCpuUsage);
         currentProcessCpuUsageSupplier = () -> invoke(processCpuUsage);
@@ -80,10 +88,6 @@ final class CpuHealthChecker implements HealthChecker {
 
     private final double targetProcessCpuLoad;
 
-    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
-
-    private static final ConcurrentHashMap<String, MethodHandle> CACHE = new ConcurrentHashMap<>();
-
     /**
      * Instantiates a new Default cpu health checker.
      *
@@ -103,18 +107,10 @@ final class CpuHealthChecker implements HealthChecker {
         this.processCpuUsageSupplier = processCpuUsageSupplier;
     }
 
-    private static double invoke(final Method method) {
-        MethodHandle mh = CACHE.get(method.getName());
+    private static double invoke(@Nullable MethodHandle mh) {
         if (mh == null) {
-            try {
-                mh = LOOKUP.unreflect(method);
-            } catch (IllegalAccessException e) {
-                throw new RuntimeException(e);
-            }
-
-            CACHE.put(method.getName(), mh);
+            return Double.NaN;
         }
-
         try {
             return (double) mh.invoke(operatingSystemBean);
         } catch (Throwable e) {
@@ -123,12 +119,19 @@ final class CpuHealthChecker implements HealthChecker {
     }
 
     @Nullable
-    private static Method detectMethod(final String name) {
+    @SuppressWarnings("ReturnValueIgnored")
+    private static MethodHandle detectMethod(final String name) {
+        if (operatingSystemBeanClass == null) {
+            return null;
+        }
         try {
             // ensure the Bean we have is actually an instance of the interface
-            requireNonNull(operatingSystemBeanClass.cast(operatingSystemBean));
-            return operatingSystemBeanClass.getMethod(name);
-        } catch (ClassCastException | NoSuchMethodException | SecurityException e) {
+            operatingSystemBeanClass.cast(operatingSystemBean);
+            final Method method = operatingSystemBeanClass.getMethod(name);
+            return LOOKUP.unreflect(method);
+        } catch (ClassCastException | NoSuchMethodException | SecurityException | IllegalAccessException e) {
+            logger.warn("Failed to detect method {}.{} for {}", operatingSystemBeanClass.getSimpleName(),
+                        name, CpuHealthChecker.class.getSimpleName(), e);
             return null;
         }
     }
@@ -153,9 +156,7 @@ final class CpuHealthChecker implements HealthChecker {
      */
     @Override
     public boolean isHealthy() {
-        final double currentSystemCpuUsage = systemCpuUsageSupplier.getAsDouble();
-        final double currentProcessCpuUsage = processCpuUsageSupplier.getAsDouble();
-        return currentSystemCpuUsage <= targetCpuUsage && currentProcessCpuUsage <= targetProcessCpuLoad;
+        return isHealthy(systemCpuUsageSupplier, processCpuUsageSupplier);
     }
 
     @VisibleForTesting

```

Bue-von-hon commented:
Hi @jrhee17 , I've addressed all of the feedback you mentioned.
I tried to be thorough in my commit message, but I'm afraid I wasn't able to do so for things I didn't understand. 🥲
Therefore, I'm going to leave a question about the changes you suggested that I didn't understand.
I look forward to your thoughtful responses. 🙏

📝 Here are my questions about the changes I didn't explain in the commit.
1. Why did you use a Methodhandle instead of a Method?
2. Why did you use the SuppressWarnings annotation to ignore the warning output?

Bue-von-hon commented:
@jrhee17  gentle ping

Bue-von-hon commented:
@jrhee17 gentle ping

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,174 @@
+    static {
+        operatingSystemBeanClass = requireNonNull(getFirstClassFound(OPERATING_SYSTEM_BEAN_CLASS_NAMES));
+        operatingSystemBean = ManagementFactory.getOperatingSystemMXBean();
+        final MethodHandle getCpuLoad = detectMethod("getCpuLoad");
+        systemCpuUsage = getCpuLoad != null ? getCpuLoad : detectMethod("getSystemCpuLoad");
+        processCpuUsage = detectMethod("getProcessCpuLoad");
```
It seems like `Usage` and `Load` are used interchangeably across this class. Can we just use `Load` for consistency? Ditto for the other field, parameter names

```suggestion
        final MethodHandle getCpuLoad = detectMethod("getCpuLoad");
        systemCpuLoad = getCpuLoad != null ? getCpuLoad : detectMethod("getSystemCpuLoad");
        processCpuLoad = detectMethod("getProcessCpuLoad");
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,174 @@
+            final Method method = operatingSystemBeanClass.getMethod(name);
+            return LOOKUP.unreflect(method);
+        } catch (ClassCastException | NoSuchMethodException | SecurityException | IllegalAccessException e) {
+            logger.warn(
+                    "Failed to detect method {}.{} for {}",
+                    operatingSystemBeanClass.getSimpleName(),
+                    name,
+                    CpuHealthChecker.class.getSimpleName(),
+                    e
+            );
```
Can you reformat these lines?

```suggestion
            logger.warn("Failed to detect method {}.{} for {}", operatingSystemBeanClass.getSimpleName(),
                        name, CpuHealthChecker.class.getSimpleName(), e);
```

jrhee17 reviewed (commented):
Looks almost done! Left some minor comments 🙇

Bue-von-hon commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    private CpuHealthChecker(double cpuUsageThreshold, double processCpuLoadThreshold,
+                             DoubleSupplier systemCpuUsageSupplier, DoubleSupplier processCpuUsageSupplier) {
```
Resolved it!

Bue-von-hon reviewed (commented):
I misunderstood your comment about testing.
I've rewritten the test correctly.
PTAL @jrhee17

Bue-von-hon commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,174 @@
+    private CpuHealthChecker(double cpuUsageThreshold, double processCpuLoadThreshold,
+                             DoubleSupplier systemCpuUsageSupplier, DoubleSupplier processCpuUsageSupplier) {
+        this.targetCpuUsage = cpuUsageThreshold;
+        this.targetProcessCpuLoad = processCpuLoadThreshold;
```
Resolved it!

Bue-von-hon commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,174 @@
+        return null;
+    }
+
+    public static CpuHealthChecker of(double targetCpuUsage, double targetProcessCpuLoad) {
```
@jrhee17 If I make this change, it will cause modifications to other files that implement this interface.
I think it would be better to address this in a new issue. If you have any other ideas, please let me know 🙂

Bue-von-hon commented:
@jrhee17 gentle ping

Bue-von-hon commented:
@jrhee17 gentle ping

Bue-von-hon commented:
@jrhee17 gentle ping

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,180 @@
+        return null;
+    }
+
+    static CpuHealthChecker of(double targetCpuUsage, double targetProcessCpuLoad) {
+        return new CpuHealthChecker(targetCpuUsage, targetProcessCpuLoad);
```
```suggestion
    static CpuHealthChecker of(double targetSystemCpuUsage, double targetProcessCpuUsage) {
        return new CpuHealthChecker(targetSystemCpuUsage, targetProcessCpuUsage);
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,180 @@
+        checkArgument(targetProcessCpuLoad >= 0 && targetProcessCpuLoad <= 1.0,
+                      "processCpuLoad: %s (expected: 0 <= processCpuLoad <= 1)", targetProcessCpuLoad);
+        this.targetSystemCpuUsage = targetSystemCpuUsage;
+        this.targetProcessCpuLoad = targetProcessCpuLoad;
```
so that others aren't confused about the difference of "load" and "usage"

```suggestion
        this.targetProcessCpuUsage = targetProcessCpuUsage;
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,180 @@
+     * @param cpuUsageThreshold the target cpu usage
+     * @param processCpuLoadThreshold the target process cpu usage
+     */
+    private CpuHealthChecker(double cpuUsageThreshold, double processCpuLoadThreshold) {
```
```suggestion
    private CpuHealthChecker(double targetSystemCpuUsage, double targetProcessCpuUsage) {
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java:
```diff
@@ -0,0 +1,180 @@
+        return null;
+    }
+
+    static CpuHealthChecker of(double targetCpuUsage, double targetProcessCpuLoad) {
```
What do you think of:
1. Moving this to `HealthChecker.java`
2. Renaming this method from `of` to `ofCpu`
3. Making this method public so that users can actually use this class
4. Adding javadocs explaining what this health checker does, and what each parameter does

jrhee17 replied:
I saw this comment late:

https://github.com/line/armeria/pull/4673#discussion_r1502153234

I'm fine if this is handled separately, I would just like to note that users won't be able to access this method since it is not `public`

jrhee17 reviewed (commented):
Sorry about the delay 😅    Looks like almost the last round for me!

ikhoon reviewed (approved):
It looks great once @jrhee17 comments are addressed.
Thanks for your effort on this work! ❤️🚀

Bue-von-hon commented:
@jrhee17 I've addressed everything you commented on.

The fact that the factory method was mentioned twice, I felt it was important. I implemented it by adding a static method to the interface. It was the best I could do, as a way to keep compatibility across subclasses as much as possible.

p.s.
I'm concerned about the class(CpuHealthChecker) being exposed outside of its intended scope. The IDE is also warning me about this.

jrhee17 reviewed (approved):
Added a minor comment mostly fixing javadocs, but looks good to me! Thanks for your patience @Bue-von-hon ! 🙇 👍 🙇

minwoox commented:
Thanks!
